Add new processor to allow controlled tag rewriting

tommasopozzetti opened the pull request:
The plugin "rewrite_tag_filter" provided alongside Fluentd allows conditional
tag rewriting to perform different processing pipelines based on log contents.
This same behavior is not achievable without it unless all logs are copied
to different output labels and then filtered through the grep plugin which
however introduces considerable overhead.
Clearly, however, the usage of this plugin needs to be constrained as it
allows to break the namespace isolation enforced by the config-reloader.

This enhancement introduces a new processor that allows to specify a new
virtual output plugin, "unique_rewrite_tag", which behaves exactly as the
"rewrite_tag_filter" (and has the same syntax and supported options), but
requires the new tag to be specified using the macro $tag(new_tag).
In order to match logs that are emitted by such output plugin it is
sufficient to specify the same macro in match and filter directives.
The processor will internally handle such configuration and ensure
that the rewritten tags are unique with respect to other namespaces and
thus will avoid breaking isolation while allowing conditional rewriting.

As an example, it might be necessary to divide logs based on the severity
which is the first word printed in the message from the app, process them
differently and then send them to the same output. This can be achieved as
follows:

```
<match kube.ns.**>
  @type unique_rewrite_tag
  <rule>
    key message
    pattern ^ERROR
    tag $tag(notifications.error)
  </rule>
  <rule>
    key message
    pattern ^FATAL
    tag $tag(notifications.fatal)
  </rule>
</match>

<filter $tag(notifications.fatal)>
  ...(perform some extra processing)
</filter>

<filter $tag(notifications.error)>
  ...(perform different processing)
</filter>

<match $tag(notifications.**)>
  ...(common output plugin)
</match>
```

jvassev commented:
Hi @tommasopozzetti,

Thanks for contributing this. Users have long requested a way to "split" a single log stream and send to two destinations but I never thought of controlled tag rewriting!

I have a few suggestions:

- Let's totally hide the fact the a rewrite_tag filter used (but keep the `invert` rule param)
- Let's call the new plugin just `retag`
- As this is an extremely useful feature it should be mentioned in the README and the FAQ (Q: How can send logs to different sinks based on log contents)
- Let's make sure the `retag` works with the `$labels` macro.
- It would be nice to let users refer to pod labels when retagging (see example)

Here is how I imagine the end result:
```
<match $labels(app=nginx)>
  @type retag
  <rule>
    key message
    pattern /^(ERROR) .*$/
    tag notifications.$app.$1 # refer to a capturing group using $number
  </rule>
</match>

<match $labels(app=apache)>
  @type retag
  <rule>
    key message
    pattern ^ERR
    tag notifications.$app.error # refer to a pod label using $alphanumeric. if label doesn exist use "null"
  </rule>
</match>


<match $tag(notification.**)> # just prepends the HASH to whatever value is found in the brackets like in your current implementation.
  @type s3
  # ...
</match>
 
Also, I plan to upgrade to fluentd 1.7.latest. Any issues with that? So, in v1.12.0 we'll have this feature and latest fluentd:)

tommasopozzetti commented:
Hi @jvassev,
Thank you for your feedback!
For what concerns hiding the `rewrite_tag_filter` plugin I agree that we can rename it to a simpler `retag`, however by looking at the [documentation](https://docs.fluentd.org/output/rewrite_tag_filter) it seems that no `remove_tag_prefix` is supported in such plugin. Also I don't see any "dangerous" option that we should drop support for. The only one is the `${tag_parts[n]}` placeholder which might become confusing given that we tend to manipulate tags and therefore change their parts. That's why I already dropped support for that in the current version.
Finally, in your example you did not use the `$tag()` macro in the tag param of the rule directive. Do you think it should be dropped? I was unsure about that as it is clearly superfluous, however I think it emphasizes that all logs coming out of this plugin need to be matched using the `$tag()` macro. I'm afraid that if it is not used in the plugin itself a common error that could arise is the following:
```
<match kube.ns.**>
  @type retag
  <rule>
    ...
    tag notifications.error
  </rule>
</match>

<match notifications.error>
  ...
</match>
```
which would not work as the `notifications.error` tag cannot be used directly. I think using the same syntax both in the `retag` plugin and in the match/filter directives that manipulate retagged logs makes it much clearer and less error-prone.

I will include support for the labels and update README and FAQ.

Finally, for what concerns Fluentd 1.7 I don't anticipate any incompatibility as the `rewrite_tag_filter` plugin should not have changed.

tommasopozzetti commented:
Actually, while reviewing the changes you proposed, I think that allowing users to refer to pod labels when retagging introduces several complexities that might not be worth the use case.

In order to allow the retag processor to know about labels it would be necessary to introduce some dependency among the two processors which I think should be avoided if possible, as it makes future enhancements more error-prone. Furthermore, given that the user is specifying in the match directive the list of labels with their value, he/she is already aware of the specific value that the label will have in that directive so he/she can directly write such value rather than the placeholder which makes everything clearer.

This does not affect the support of the `$labels` macro in the match directive of the `retag` plugin, which is supported as with any other plugin.

I have also renamed the plugin as `retag` and added a section in the README and FAQ for this new plugin.

Please let me know your thoughts on this

jvassev commented:
Hi, thanks for having a closer look!
You are right about the labels - not worth the complexity.

I would still go for the simpler syntax in the `<rule>`:
```
<match kube.ns.**>
  @type retag
  <rule>
    ...
    tag notifications.error
  </rule>
</match>

<match $tag(notifications.error)>
  ...
</match>
```

This way, the new $tag macro is used only in the `<match>` - symmetric to `$labels`. Like the `$labels` macro, $tag lets you refer to data in a safe way. It just happens that the data for `$labels` is produced transparently while the data for `$tags` is produced explicitly using the new `retag` plugin.

The syntax: 
```
<match notifications.error>
  ...
</match>
```
will be rejected by the rule which enforces same-namespace (https://github.com/vmware/kube-fluentd-operator/blob/master/config-reloader/processors/thisns.go#L23).

On the topic of `thisns`, also for the implementation of:
```
func (p *uniqueRewriteTagState) createUniqueTag(tag, namespace string) string {
	return util.Hash(namespace, "") + "." + tag
}
```

I think it is safer to preserve the `kube.{namespace}.` prefix and generate the new  tag like this:
```
func (p *uniqueRewriteTagState) createUniqueTag(tag, namespace string) string {
	return "kube."+ namespace + "._retag." + tag
}
```
This way you can still use `<match **>`  to capture retagged events. Also, it is symmetric to the the labels processor.

tommasopozzetti commented:
All edits should be ready!
I had thought about the `kube.{namespace}` prefix as well but I was unsure about it as it might lead to the confusing case of someone specifying precisely the `<match kube.{namespace}.**>` directive and having no reason to believe that such directive would capture the `notifications.error` stream coming out of the `retag` plugin, which is however the case.
Nevertheless, this is probably an edge case and it does make sense to have the `<match **>` directive be able to match also outputs from the `retag` plugin, so I did include this change.

tommasopozzetti commented:
Hi @jvassev, is there anything left that I can do before we can merge this in?

carpenterm commented:
@tommasopozzetti apologies for the delay, any chance you could rebase and we'll get this merged?

tommasopozzetti commented:
Hi @carpenterm, thank you for taking a look.
I have rebased the PR to solve conflicts.

dimalbaby commented:
@tommasopozzetti Thanks for the PR
